JS UI Thermostat set point onclick fix.

ajarzyn opened the pull request:
Only bigtext was updated. Updating onclick function parameter.

On update_device only bigtext is updated.
`ShowSetpointPopup(event, 40, false, 1.0)` data value stayed the same thus after opening set point again value from page load was loaded.

```
<td id="bigtext">1.0° C</td>
<td id="img"><img src="images/override.png" class="lcursor" onclick="ShowSetpointPopup(event, 40, false, 1.0);" width="48" height="48"></td>
```

Change only for thermostat device.

PS.
I was trying to use following line:
UtilityController.js:814: `'<img src="images/' + item.Image + '" class="lcursor" onclick="ShowSetpointPopup(event, ' + item.idx + ', ' + item.Protected + ', ' + item.Data + ');" height="48" width="48" ></td>\n';` in place of regex to update img but unfortunately in `RefreshItem = function (item) {` `item` does not contain `Image` member thus regex.

It would be good to have initial definition of img field (line814) in one place not to double the code, but using regex here isn't (in my opinion) most efficient solution.

Could someone look into why there is not `Image` member?

ajarzyn committed:
JS UI Thermostat set point onclick fix.
Only bigtext was updated. Updating onclick function parameter.

gizmocuz commented:
The "Image" field is only set when the user chooses a custom icon over the default one.
If none is provided, the default icon is set via the JS script.
In this case, there is also no override48_On.png as there is for the custom icons (s they have a on/off and small icon)

Happy to merge your PR if you don't see issues doing so

ajarzyn commented:
> Happy to merge your PR if you don't see issues doing so

I'm sorry but I do not understand this sentence. Is it politeness, good manners or real question?
I mean I would never made a PR if I wouldn't like it to be merged :)

About Images I do understand that it makes less data to be throw in the requests and that's fine.
With `data['Image']` it would be also 'no go' without refactoring styling the controls.
I've asked about it only because I wasn't sure that we should use regex/html parsing/string operations, but if you feel this is proper way to go please feel free and merge.

Best to you @gizmocuz and thank you for you good work.

ajarzyn commented on www/app/UtilityController.js:
```diff
@@ -433,6 +433,8 @@ define(['app', 'livesocket'], function (app) {
 				else if ((item.Type == "Thermostat") && (item.SubType == "SetPoint")) {
 					status = "";
 					bigtext = item.Data + '\u00B0 ' + $scope.config.TempSign;
+					const regex = /ShowSetpointPopup\(event, \d*, \w*, (\d+\.?)*\d+\);/gm;
```
One last thought.
What if signature for ShowSetpointPopup will change?
Shouldn't we make it depended on the JS code we have?

ajarzyn replied:
> Do you also perform this action on the dashboard?

This proves my point, now I know about 3 (4?) different places where this is set.
- initialization of the web (dashboard)
- `refreshItem` (dashboard)
- During initialization of the web (Utils)
- `refreshItem` (Utils)

I feel like it should be wrapped into a class or some structure.
I would like to propose another PR, to parametrize this set point (I've already rose this in different [thread](https://github.com/domoticz/domoticz/issues/4055#issuecomment-993254561))

Step/limits would be useful.

ajarzyn commented on www/app/UtilityController.js:
```diff
@@ -433,6 +433,8 @@ define(['app', 'livesocket'], function (app) {
 					status = "";
 					bigtext = item.Data + '\u00B0 ' + $scope.config.TempSign;
+					const regex = /ShowSetpointPopup\(event, \d*, \w*, (\d+\.?)*\d+\);/gm;
+					img = $(id + " #img").html().replace(regex, 'ShowSetpointPopup(event, ' + item.idx + ', ' + item.Protected + ', ' + item.Data + ');')
```
Additional 'last thought'
wouldn't it be better to update only attribute for onClick?
Something like this (not tested yet):
`$(id + " #img").attr('ShowSetpointPopup(event, ' + item.idx + ', ' + item.Protected + ', ' + item.Data + ')')`

This way we do not need to use regex and we update this parameter directly under this if, not waiting for the later if with `img != ''`

Problem with doubling function signature will still exist.

ajarzyn replied:
Changed to attribute handling.

ajarzyn reviewed (commented):
Please check additional comments.

gizmocuz commented:
The reason for asking is because you seem to have some doubts about using a regex. It did work nicely and my JS regex experience is limited.
I think the double function will still exist and updating only the attribute might seems better
Both work, I leave this up to you 😉

ajarzyn commented:
Please merge it. If I will find time (as I have some more debugging to do) related to Thermostat python plugin NameType. 
I will create another PR with only onclick attr set - as I think this is 'better' approach.

gizmocuz commented:
Thanks I will, just two more quick questions
Does the regex now have to be in quotes '<regex>', i always put strings in quotes
Do you also perform this action on the dashboard?

ajarzyn committed:
Thermostat set point onlick update.

Remove regex. Change mothod to set attribute for DOM object. Apply change for dashboard.

ajarzyn committed:
Thermostat setpoint - change spaces to tabs.

ajarzyn commented:
ad 1. No it's just the language notation [stack answer](https://stackoverflow.com/questions/31491908/why-are-regular-expression-strings-not-encapsulated-in-quotes-in-javascript/31491967#31491967)
ad 2. No I haven't but I created another commit and hopefully it will be all good now: [5d706f4](https://github.com/domoticz/domoticz/pull/5297/commits/5d706f451f87c6afed31362e9877725bf2f65541) and corrected whitespaces in the next one: [04283e6](https://github.com/domoticz/domoticz/pull/5297/commits/04283e63a79d177c7b498a237891bf7457486138)

Please check my comment in the conversation.

ajarzyn committed:
Fix thermostat set point.

Reverted last commit for UtilityController, regex removed, update of a attr for thermostat item added.

ajarzyn commented:
I wanted to write that I tested it (it's late... almost 2AM), so I did one more check and it appeared that I've edited wrong line of thermostat code.

This time [23a79a9](https://github.com/domoticz/domoticz/pull/5297/commits/23a79a91e8514966ee1d1d8327523d35a1de0626) solution is fully tested, for Utilities and Dashboad.

Sorry for the mess in commits.

gizmocuz commented:
Thank you very much for the PR and your time! Looks great !
